Selection model and selection architecture refactor

clairetang6 opened the pull request:
This PR is an overhaul of the selection architecture. The goal is to clarify the roles of each piece of the selection machinery and also to create a new models to represent selections and inspections. 

There's more information about the goals of the selection architecture refactor and notes on the selection machinery on [this page](https://github.com/bokeh/bokeh/wiki/Selection-Architecture-Refactor) in the wiki. 

- [x] issues: fixes #6845 #7416 #7054
- [ ] tests added / passed
- [ ] release document entry (if new feature or API change)

mattpap commented:
Extremely bad timing given PR #7233. CoffeeScript bits will need to be rewritten in TypeScript, on top of that PR.

mattpap commented on bokehjs/src/coffee/core/selection_manager.ts:
```diff
@@ -10,47 +10,152 @@ import {DataSource} from "models/sources/data_source"
+
+  hit_test(geometry: Geometry, renderer_views: RendererView[]): HitTestResult {
+    let hit_test_result_renderers = []
+    for (let r of renderer_views) {
```
Both variables (and a few more) can be `const`. The preference should be to use `const`, and use `let` only when a variable is meant to be reassigned (e.g. a loop counter). `const` doesn't relate to the mutability or immutability of the object behind a variable, so you can push to `const arr = []`.

clairetang6 replied:
Ok, I understand using `const` for arrays and objects now, but is the `let r of renderer_views` not being reassigned?

mattpap replied:
Right, `for (const r of arr)` may seem a bit unintuitive, but technically there is no explicit assignment to `r` (so, `r = ...`) anywhere, so you can safely use `const`. If it was `for (i = 0, ...)` or `while(r = arr.unshift()) ...` then `let` would be a must.

mattpap commented on bokehjs/src/coffee/core/selection_manager.ts:
```diff
@@ -10,47 +10,152 @@ import {DataSource} from "models/sources/data_source"
+
+export abstract class SelectionPolicy extends Model {
+
+  hit_test(geometry: Geometry, renderer_views: RendererView[]): HitTestResult{
```
Given `SelectionPolicy` is an abstract class, this method would be better defined as abstract as well.

mattpap commented on bokehjs/src/coffee/core/selection_manager.ts:
```diff
@@ -10,47 +10,152 @@ import {DataSource} from "models/sources/data_source"
+    return null
+  }
+
+  do_selection(hit_test_result: HitTestResult | null, renderer_views: RendererView[], final: boolean, append: boolean): boolean {
```
`HitTestResult` is defined as `Selection | null`, so this additional `null` is unnecessary.

mattpap commented on bokehjs/src/coffee/core/selection_manager.ts:
```diff
@@ -10,47 +10,152 @@ import {DataSource} from "models/sources/data_source"
+    const glyph_renderer_views: RendererView[] = []
+    const graph_renderer_views: RendererView[] = []
+    for (let r of renderer_views) {
+      if (r.model.type == 'GlyphRenderer'){
```
Is there a reason to use `type`? The right solution is to use `instanceof`, because this way you respect the fact that models can be extended. `type` should be used only for serialization purpose and extremely rare occasions when nothing else works (e.g. mutually recursive imports that result in infinite recursion).

clairetang6 replied:
I just tried this again, and think I may have originally tried it and then run into this issue where ColumnarDataSource imports SelectionManager which imports GlyphRenderer which imports RemoteDataSource which breaks. This is maybe because RemoteDataSource inherits from ColumnarDataSource?

mattpap commented on bokehjs/src/coffee/models/selections/selection.ts:
```diff
@@ -0,0 +1,57 @@
+  [key: string]: any
+
+  initialize(options: any): void {
+    super.initialize(options)
```
Missing `attrs`. Compiler is actually complaining about this. This is such a common error across bokehjs (views take only `options` argument).

clairetang6 commented:
@mattpap, I'm getting the error "Unable to compile TypeScript" when I try to run JS tests. I just tried on master after a fresh npm install and am getting the same error. Do you know what the problem might be?

mattpap commented:
> (...) Unable to compile TypeScript (...)

@clairetang6, it would be useful to know the full output. However, we recently switched to node 8 (npm 5), so this may be the reason. Update your installation (node 8 is available from conda-forge channel), remove node_modules and reinstall and rebuild.

clairetang6 commented:
@mattpap, I updated to node v8.9.3 and npm 5.6.0, and did a fresh `npm install`, but running `gulp test` still is not working. Building with `gulp build` works. It's just `gulp test` that is giving me an error. 

```
(bokeh) ClairesMacBook:bokehjs claire$ gulp test
[15:52:20] Using gulpfile ~/Documents/bokeh/bokehjs/gulpfile.js
[15:52:20] Starting 'defaults:generate'...
[15:52:22] generate_defaults.py: /Users/claire/Documents/bokeh/bokeh/util/deprecation.py:34: BokehDeprecationWarning: DynamicImageRenderer was deprecated in Bokeh 0.12.7 and will be removed, use GeoViews for GIS functions on top of Bokeh (http://geo.holoviews.org) instead.
  warn(message)

[15:52:22] generate_defaults.py: Wrote ./test/.generated_defaults/models_defaults.ts with 149 model classes
[15:52:22] generate_defaults.py: Wrote ./test/.generated_defaults/widgets_defaults.ts with 22 model classes
[15:52:22] Finished 'defaults:generate' after 1.5 s
[15:52:22] Starting 'test'...
⨯ Unable to compile TypeScript:
test/core/has_props.ts (10,25): Type '(attributes: any, options: any) => void' is not a constructor function type. (2507)
test/core/has_props.ts (11,21): Property 'type' does not exist on type 'TestModel'. (2339)
test/core/has_props.ts (13,33): Type '(attributes: any, options: any) => void' is not a constructor function type. (2507)
test/core/has_props.ts (17,29): Property 'type' does not exist on type 'SubclassWithProps'. (2339)
test/core/has_props.ts (18,19): Property 'define' does not exist on type 'typeof SubclassWithProps'. (2339)
test/core/has_props.ts (26,32): Property 'type' does not exist on type 'SubSubclassWithProps'. (2339)
test/core/has_props.ts (27,22): Property 'define' does not exist on type 'typeof SubSubclassWithProps'. (2339)
test/core/has_props.ts (31,34): Type '(attributes: any, options: any) => void' is not a constructor function type. (2507)
test/core/has_props.ts (32,30): Property 'type' does not exist on type 'SubclassWithMixins'. (2339)
test/core/has_props.ts (33,20): Property 'mixin' does not exist on type 'typeof SubclassWithMixins'. (2339)
test/core/has_props.ts (36,33): Property 'type' does not exist on type 'SubSubclassWithMixins'. (2339)
test/core/has_props.ts (37,23): Property 'mixin' does not exist on type 'typeof SubSubclassWithMixins'. (2339)
test/core/has_props.ts (39,42): Type '(attributes: any, options: any) => void' is not a constructor function type. (2507)
test/core/has_props.ts (40,38): Property 'type' does not exist on type 'SubclassWithMultipleMixins'. (2339)
test/core/has_props.ts (41,28): Property 'mixin' does not exist on type 'typeof SubclassWithMultipleMixins'. (2339)
test/core/has_props.ts (43,38): Type '(attributes: any, options: any) => void' is not a constructor function type. (2507)
test/core/has_props.ts (47,34): Property 'type' does not exist on type 'SubclassWithNumberSpec'. (2339)
test/core/has_props.ts (48,24): Property 'define' does not exist on type 'typeof SubclassWithNumberSpec'. (2339)
test/core/has_props.ts (53,40): Type '(attributes: any, options: any) => void' is not a constructor function type. (2507)
test/core/has_props.ts (57,36): Property 'type' does not exist on type 'SubclassWithDistanceSpec'. (2339)
test/core/has_props.ts (58,26): Property 'define' does not exist on type 'typeof SubclassWithDistanceSpec'. (2339)
test/core/has_props.ts (63,41): Type '(attributes: any, options: any) => void' is not a constructor function type. (2507)
test/core/has_props.ts (67,37): Property 'type' does not exist on type 'SubclassWithTransformSpec'. (2339)
test/core/has_props.ts (68,27): Property 'define' does not exist on type 'typeof SubclassWithTransformSpec'. (2339)
test/core/has_props.ts (73,40): Type '(attributes: any, options: any) => void' is not a constructor function type. (2507)
test/core/has_props.ts (78,36): Property 'type' does not exist on type 'SubclassWithOptionalSpec'. (2339)
test/core/has_props.ts (79,26): Property 'define' does not exist on type 'typeof SubclassWithOptionalSpec'. (2339)
test/core/has_props.ts (98,23): Property 'properties' does not exist on type 'TestModel'. (2339)
test/core/has_props.ts (99,23): Property 'attributes' does not exist on type 'TestModel'. (2339)
test/core/has_props.ts (104,23): Property 'properties' does not exist on type 'SubclassWithProps'. (2339)
test/core/has_props.ts (109,23): Property 'properties' does not exist on type 'SubSubclassWithProps'. (2339)
test/core/has_props.ts (115,23): Property 'properties' does not exist on type 'SubclassWithMixins'. (2339)
test/core/has_props.ts (121,23): Property 'properties' does not exist on type 'SubSubclassWithMixins'. (2339)
test/core/has_props.ts (127,23): Property 'properties' does not exist on type 'SubclassWithMultipleMixins'. (2339)
test/core/has_props.ts (133,17): Only a void function can be called with the 'new' keyword. (2350)
test/core/has_props.ts (133,17): Expected 0 arguments, but got 1. (2554)
test/core/has_props.ts (135,24): Property 'materialize_dataspecs' does not exist on type 'SubclassWithNumberSpec'. (2339)
test/core/has_props.ts (140,17): Only a void function can be called with the 'new' keyword. (2350)
test/core/has_props.ts (140,17): Expected 0 arguments, but got 1. (2554)
test/core/has_props.ts (143,24): Property 'materialize_dataspecs' does not exist on type 'SubclassWithNumberSpec'. (2339)
test/core/has_props.ts (148,17): Only a void function can be called with the 'new' keyword. (2350)
test/core/has_props.ts (148,17): Expected 0 arguments, but got 1. (2554)
test/core/has_props.ts (151,25): Property 'materialize_dataspecs' does not exist on type 'SubclassWithDistanceSpec'. (2339)
test/core/has_props.ts (155,25): Property 'materialize_dataspecs' does not exist on type 'SubclassWithDistanceSpec'. (2339)
test/core/has_props.ts (160,17): Only a void function can be called with the 'new' keyword. (2350)
test/core/has_props.ts (160,17): Expected 0 arguments, but got 1. (2554)
test/core/has_props.ts (162,11): Property 'properties' does not exist on type 'SubclassWithOptionalSpec'. (2339)
test/core/has_props.ts (163,24): Property 'materialize_dataspecs' does not exist on type 'SubclassWithOptionalSpec'. (2339)
[15:52:25] 'test' errored after 2.95 s
[15:52:25] Error in plugin 'mocha'
Message:
    tests failed
```

mattpap commented:
@clairetang6, lets start from updating this branch with most recent `master`.

clairetang6 commented:
@mattpap, I have been trying to run gulp test from the master branch and that’s where I’m getting the error, so there must be something wrong with my dev setup, but I don’t know what it is. Are there any globally installed node packages that aren’t specified? I checked that my versions of gulp and typescript are up-to-date.

mattpap commented:
> I have been trying to run gulp test from the master branch and that’s where I’m getting the error (...)

I see. Hm. The error messages indicate that there's not enough type information to compile has_props' tests. I presume that by "fresh install" you mean `rm -rf node_modules build; npm install; gulp build`? If so, did you ever install anything with `npm install -g` (except `npm` itself)? Maybe there are some outdated global packages being picked up. At this point I don't have other ideas. I already disabled all caching and what not, to avoid situations like this one. Maybe you could send me output from `npm ls` (e.g. as a gist), to compare what versions were installed on your system. You could also remove `/tmp/ts-node*` directories, to make sure that no caches are being picked up.

clairetang6 commented:
@mattpap, yeah that is what I meant by fresh install. 

I got it to work after completely cleaning the directories with `git clean -xdf`. It must have been some caching issue. Thank you for your help!

clairetang6 commented:
@mattpap, I have a lot of TypeScript errors that I'm not sure how to resolve. A lot of them look like: 
```
[21:50:07] src/coffee/models/markers/marker.ts(132,3): error TS4053: Return type of public method from exported class has or is using name 'Selection' from external module "/Users/claire/Documents/bokeh/bokehjs/src/coffee/models/selections/selection" but cannot be named.
```
or 
```
[21:50:04] src/coffee/models/selections/selection.ts(98,53): error TS2339: Property 'indices' does not exist on type 'Selection'.
```

mattpap commented:
> error TS4053: Return type of public method from exported class has or is using name (...)

If you override a member of a parent class from one module in another one, then you have to provide all the necessary types. This wouldn't be normally required, but we use auto-generated type declarations. Consider an example:

```typescript
// mod1.ts
export class A {}
export class X {
  func(): A { return new A() }
}
// mod2.ts
import {X} from "./mod1"
export class Y extends X {
  func() {
    console.log("override")
    return super.func()
  }
}
```
This will result in an error, because for `Y.func`'s declaration to be generated, `A` must be known in `mod2.ts`. So you have to at least import `A` in `mod2.ts`, but preferably also explicitly type the return `Y.func` to make all this more straightforward to the reader. This is why we have a (so far not written down) policy to always provide explicit return types for all functions and methods.

mattpap commented:
> Property 'indices' does not exist on type 'Selection'.

This is because whenever you use `this.define()` or `this.internal()` to define model's properties, this only affects the runtime behavior of a model, but not its type/static behavior. For that you have to provide appropriate class attributes. So, for `Selection` model this will be (more or less):

```typescript
export class Selection extends Model {
  static initClass() {
    this.prototype.type = "Selection";

    this.define({
      indices:         [ p.Array,   [] ],
    });

    this.internal({
      final:             [ p.Boolean     ],
      line_indices:      [ p.Array,   [] ],
      selected_glyphs:   [ p.Array,   [] ],
      get_view:          [ p.Any         ],
      multiline_indices: [ p.Any,     {} ],
    });
  }

  indices: number[]

  final: boolean
  line_indices: number[]
  selected_glyphs: Glyph[]
  get_view:  () => GlyphView
  multiline_indices: {[key: string]: number[]} 
```
Note those types are an approximation, so change those to whatever is actually needed. Similar approach should be applied to all new models and all new or changed properties of existing models.

mattpap commented on bokehjs/src/coffee/models/renderers/graph_renderer.ts:
```diff
@@ -17,34 +22,34 @@ export class GraphRendererView extends RendererView {
+  set_data(request_render: boolean) {
     // TODO (bev) this is a bit clunky, need to make sure glyphs use the correct ranges when they call
     // mapping functions on the base Renderer class
     if (request_render == null) { request_render = true; }
```
Instead of sprinkling `true` around the code, this line could be removed and the signature changed to `set_data(request_render: boolean = true)`. However, I'm not sure why all those changes are needed in this PR. If those changes aren't strictly necessary, you are greatly reducing mergeability of this PR this way.

mattpap replied:
Actually, I already fixed this in `master`.

bryevdv commented:
@clairetang6 should this be marked `ready` now?

clairetang6 commented on bokehjs/src/coffee/models/selections/selection.ts:
```diff
@@ -0,0 +1,115 @@
+
+  get selected_glyph(): Glyph | null {
+    if(this.selected_glyphs.length > 0)
+      return this.selected_glyphs[0]
```
The goal here is to eventually be able to get any of the glyphs that were hit as part of the selection. For now, this is just used as an implementation detail to let a Line GlyphRenderer know its Line was selected.

bryevdv commented:
@mattpap I'd like to look at this today. Can you fix up the merge conflicts from your other PR?

mattpap commented:
@bryevdv, I would first merge PR #7467, because it will bring merge conflicts in almost the same places.

bryevdv commented:
@mattpap OK, I have merged the other PR. Please fix up this one ASAP, I'd like to review and merge this today.

mattpap commented:
> OK, I have merged the other PR. Please fix up this one ASAP, I'd like to review and merge this today.

I will do it in a moment, however mind that edit tools need to be updated to the new API.

bryevdv commented:
> I will do it in a moment, however mind that edit tools need to be updated to the new API. 

please ping @philippjfr on a new issue

philippjfr commented:
Right, happy to help out on that. I'll be fairly busy until tomorrow afternoon though.

mattpap commented:
I resolved merge conflicts. There however a few errors to fix:
```
[17:42:16] src/coffee/core/selection_manager.ts(82,5): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/graphs/graph_hit_test_policy.ts(58,5): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/graphs/graph_hit_test_policy.ts(61,13): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/graphs/graph_hit_test_policy.ts(92,22): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/graphs/graph_hit_test_policy.ts(116,5): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/graphs/graph_hit_test_policy.ts(120,5): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/graphs/graph_hit_test_policy.ts(124,13): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/graphs/graph_hit_test_policy.ts(160,22): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/graphs/graph_hit_test_policy.ts(180,5): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/graphs/graph_hit_test_policy.ts(184,5): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/graphs/graph_hit_test_policy.ts(188,13): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/plots/plot_canvas.ts(383,9): error TS2554: Expected 1 arguments, but got 0.
[17:42:16] src/coffee/models/selections/interaction_policy.ts(15,7): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/selections/interaction_policy.ts(17,15): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/tools/edit/box_edit_tool.ts(127,7): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/tools/edit/edit_tool.ts(90,27): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/tools/edit/point_draw_tool.ts(79,7): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/tools/edit/poly_draw_tool.ts(161,27): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/tools/edit/poly_draw_tool.ts(184,7): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/tools/edit/poly_edit_tool.ts(34,21): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/tools/edit/poly_edit_tool.ts(36,9): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/tools/edit/poly_edit_tool.ts(42,9): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/tools/edit/poly_edit_tool.ts(66,19): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/tools/edit/poly_edit_tool.ts(88,5): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/tools/edit/poly_edit_tool.ts(103,21): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/tools/edit/poly_edit_tool.ts(122,21): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/tools/edit/poly_edit_tool.ts(123,7): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/tools/edit/poly_edit_tool.ts(151,19): error TS2531: Object is possibly 'null'.
[17:42:16] src/coffee/models/tools/edit/poly_edit_tool.ts(175,5): error TS2531: Object is possibly 'null'.
```

bryevdv commented:
In addition to the above compilation errors, https://github.com/bokeh/bokeh/pull/7479 has an integration test that will need to be updated to set the selection programmatically in the new way. 

@clairetang6 apologies it has taken me so long to evaluate this PR. Once the above issues are fixed up, we should be good. I've asked to pause further TS port updates this cycle, so that this can be the last major PR to go in before some extensive testing.

philippjfr commented:
What's the best way for me to start resolving issues with the edit tools here? Should I just start pushing fixes? I'd be happy to start on that once it has been rebased again.

clairetang6 commented:
@bryevdv @philippjfr, I merged master before the last commit and updated everything in the new edit tools. I tested the example scripts for the edit tools and everything seems to be working. I also updated the selection_nonselection example that @mattpap just added.

mattpap commented on bokehjs/src/coffee/models/renderers/glyph_renderer.ts:
```diff
@@ -176,37 +179,30 @@ export class GlyphRendererView extends RendererView {
-    let selected: number[]
-    if (!_selected || (_selected.length === 0)) {
-      selected = [];
+    let { selected } = this.model.data_source;
```
`const`

mattpap commented on bokehjs/src/coffee/models/renderers/glyph_renderer.ts:
```diff
@@ -176,37 +179,30 @@ export class GlyphRendererView extends RendererView {
-    if (!_selected || (_selected.length === 0)) {
-      selected = [];
+    let { selected } = this.model.data_source;
+    let selected_full_indices: number[] = []
```
Assignment here is unnecessary, as you are doing this anyway in the next block of code.

mattpap commented on bokehjs/src/coffee/models/selections/interaction_policy.ts:
```diff
@@ -0,0 +1,68 @@
+import {Geometry} from "core/geometry"
+import {HitTestResult} from "core/hittest"
+import {GlyphRendererView} from "models/renderers/glyph_renderer"
+import {ColumnarDataSource} from "models/sources/columnar_data_source"
```
Use relative imports when referring from models to models.

mattpap commented on bokehjs/src/coffee/models/sources/data_source.ts:
```diff
@@ -24,11 +24,19 @@ export abstract class DataSource extends Model {
+
+    if (!this.selected) {
+      this.selected = new Selection()
+    }
```
Why not make this the default value for `selected` property?

clairetang6 replied:
I also wasn't able to set this default on the python side. If I add the default of `lambda: Selection`, I get the following error:

```
Traceback (most recent call last):
  File "selection_model.py", line 1, in <module>
    from bokeh.plotting import figure, show
  File "/Users/claire/Documents/bokeh/bokeh/plotting/__init__.py", line 4, in <module>
    from ..models import ColumnDataSource; ColumnDataSource
  File "/Users/claire/Documents/bokeh/bokeh/models/__init__.py", line 16, in <module>
    from .annotations import *
  File "/Users/claire/Documents/bokeh/bokeh/models/annotations.py", line 16, in <module>
    from ..core.property_mixins import FillProps, LineProps, TextProps
  File "/Users/claire/Documents/bokeh/bokeh/core/property_mixins.py", line 71, in <module>
    class FillProps(HasProps):
  File "/Users/claire/Documents/bokeh/bokeh/core/property_mixins.py", line 79, in FillProps
    fill_alpha = NumberSpec(default=1.0, accept_datetime=False, help=_alpha_help % "fill paths")
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 1595, in __init__
    super(NumberSpec, self).__init__(key_type, Float, default=default, help=help)
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 1531, in __init__
    help=help
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 653, in __init__
    super(Either, self).__init__(default=default, help=help)
  File "/Users/claire/Documents/bokeh/bokeh/core/property/bases.py", line 65, in __init__
    self.validate(self._raw_default())
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 690, in validate
    if not (value is None or any(param.is_valid(value) for param in self.type_params)):
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 690, in <genexpr>
    if not (value is None or any(param.is_valid(value) for param in self.type_params)):
  File "/Users/claire/Documents/bokeh/bokeh/core/property/bases.py", line 259, in is_valid
    self.validate(value)
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 1299, in validate
    raise ValueError("expected an element of %s, got %r" % (self, value))
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 1279, in __str__
    return "%s(%s, %s)" % (self.__class__.__name__, self.keys_type, self.values_type)
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 663, in __str__
    return "%s(%s)" % (self.__class__.__name__, ", ".join(map(str, self.type_params)))
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 435, in __str__
    return "%s(%s)" % (self.__class__.__name__, self.instance_type.__name__)
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 445, in instance_type
    self._instance_type = getattr(import_module(module, "bokeh"), name)
  File "/anaconda/envs/bokeh/lib/python3.5/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/Users/claire/Documents/bokeh/bokeh/models/transforms.py", line 17, in <module>
    from .sources import ColumnarDataSource
  File "/Users/claire/Documents/bokeh/bokeh/models/sources.py", line 19, in <module>
    class DataSource(Model):
  File "/Users/claire/Documents/bokeh/bokeh/models/sources.py", line 26, in DataSource
    """)
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 432, in __init__
    super(Instance, self).__init__(default=default, help=help)
  File "/Users/claire/Documents/bokeh/bokeh/core/property/bases.py", line 65, in __init__
    self.validate(self._raw_default())
  File "/Users/claire/Documents/bokeh/bokeh/core/property/bases.py", line 133, in _raw_default
    return self._copy_default(self._default)
  File "/Users/claire/Documents/bokeh/bokeh/core/property/bases.py", line 123, in _copy_default
    return default()
  File "/Users/claire/Documents/bokeh/bokeh/models/sources.py", line 24, in <lambda>
    selected = Instance(Selection, lambda: Selection(), help="""
  File "/Users/claire/Documents/bokeh/bokeh/model.py", line 220, in __init__
    default_theme.apply_to_model(self)
  File "/Users/claire/Documents/bokeh/bokeh/themes/theme.py", line 158, in apply_to_model
    model.apply_theme(self._for_class(model.__class__))
  File "/Users/claire/Documents/bokeh/bokeh/themes/theme.py", line 143, in _for_class
    self._add_glyph_defaults(base, combined)
  File "/Users/claire/Documents/bokeh/bokeh/themes/theme.py", line 126, in _add_glyph_defaults
    from ..models.glyphs import Glyph
  File "/Users/claire/Documents/bokeh/bokeh/models/glyphs.py", line 44, in <module>
    from ..core.property_mixins import FillProps, LineProps, ScalarFillProps, ScalarLineProps, TextProps
ImportError: cannot import name 'FillProps'
```

mattpap replied:
Well, this is something that can be dealt with, but nothing we should be concerned with in this PR.

mattpap commented on bokehjs/src/coffee/models/sources/columnar_data_source.ts:
```diff
@@ -58,6 +65,10 @@ export abstract class ColumnarDataSource extends DataSource {
+
+    if (!this.selection_policy)
+      this.selection_policy = new UnionRenderers() // added here instead of as default because can't set default
+                                                   // on Python side without error
```
> on Python side without error

What error? Perhaps this is something we could fix, to allow better management of default values.

clairetang6 replied:
Similar error to above: 

```
Traceback (most recent call last):
  File "selection_model.py", line 1, in <module>
    from bokeh.plotting import figure, show
  File "/Users/claire/Documents/bokeh/bokeh/plotting/__init__.py", line 4, in <module>
    from ..models import ColumnDataSource; ColumnDataSource
  File "/Users/claire/Documents/bokeh/bokeh/models/__init__.py", line 16, in <module>
    from .annotations import *
  File "/Users/claire/Documents/bokeh/bokeh/models/annotations.py", line 16, in <module>
    from ..core.property_mixins import FillProps, LineProps, TextProps
  File "/Users/claire/Documents/bokeh/bokeh/core/property_mixins.py", line 71, in <module>
    class FillProps(HasProps):
  File "/Users/claire/Documents/bokeh/bokeh/core/property_mixins.py", line 79, in FillProps
    fill_alpha = NumberSpec(default=1.0, accept_datetime=False, help=_alpha_help % "fill paths")
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 1595, in __init__
    super(NumberSpec, self).__init__(key_type, Float, default=default, help=help)
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 1531, in __init__
    help=help
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 653, in __init__
    super(Either, self).__init__(default=default, help=help)
  File "/Users/claire/Documents/bokeh/bokeh/core/property/bases.py", line 65, in __init__
    self.validate(self._raw_default())
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 690, in validate
    if not (value is None or any(param.is_valid(value) for param in self.type_params)):
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 690, in <genexpr>
    if not (value is None or any(param.is_valid(value) for param in self.type_params)):
  File "/Users/claire/Documents/bokeh/bokeh/core/property/bases.py", line 259, in is_valid
    self.validate(value)
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 1299, in validate
    raise ValueError("expected an element of %s, got %r" % (self, value))
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 1279, in __str__
    return "%s(%s, %s)" % (self.__class__.__name__, self.keys_type, self.values_type)
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 663, in __str__
    return "%s(%s)" % (self.__class__.__name__, ", ".join(map(str, self.type_params)))
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 435, in __str__
    return "%s(%s)" % (self.__class__.__name__, self.instance_type.__name__)
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 445, in instance_type
    self._instance_type = getattr(import_module(module, "bokeh"), name)
  File "/anaconda/envs/bokeh/lib/python3.5/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/Users/claire/Documents/bokeh/bokeh/models/transforms.py", line 17, in <module>
    from .sources import ColumnarDataSource
  File "/Users/claire/Documents/bokeh/bokeh/models/sources.py", line 33, in <module>
    class ColumnarDataSource(DataSource):
  File "/Users/claire/Documents/bokeh/bokeh/models/sources.py", line 45, in ColumnarDataSource
    """)
  File "/Users/claire/Documents/bokeh/bokeh/core/properties.py", line 432, in __init__
    super(Instance, self).__init__(default=default, help=help)
  File "/Users/claire/Documents/bokeh/bokeh/core/property/bases.py", line 65, in __init__
    self.validate(self._raw_default())
  File "/Users/claire/Documents/bokeh/bokeh/core/property/bases.py", line 133, in _raw_default
    return self._copy_default(self._default)
  File "/Users/claire/Documents/bokeh/bokeh/core/property/bases.py", line 123, in _copy_default
    return default()
  File "/Users/claire/Documents/bokeh/bokeh/models/sources.py", line 43, in <lambda>
    selection_policy = Instance(SelectionPolicy, lambda: IntersectRenderers(), help="""
  File "/Users/claire/Documents/bokeh/bokeh/model.py", line 220, in __init__
    default_theme.apply_to_model(self)
  File "/Users/claire/Documents/bokeh/bokeh/themes/theme.py", line 158, in apply_to_model
    model.apply_theme(self._for_class(model.__class__))
  File "/Users/claire/Documents/bokeh/bokeh/themes/theme.py", line 143, in _for_class
    self._add_glyph_defaults(base, combined)
  File "/Users/claire/Documents/bokeh/bokeh/themes/theme.py", line 126, in _add_glyph_defaults
    from ..models.glyphs import Glyph
  File "/Users/claire/Documents/bokeh/bokeh/models/glyphs.py", line 44, in <module>
    from ..core.property_mixins import FillProps, LineProps, ScalarFillProps, ScalarLineProps, TextProps
ImportError: cannot import name 'FillProps'
```

mattpap commented on bokehjs/src/coffee/models/sources/columnar_data_source.ts:
```diff
@@ -32,6 +34,10 @@ export abstract class ColumnarDataSource extends DataSource {
 
+  inspected: Selection
+  selection_policy: SelectionPolicy
+  selection_manager: SelectionManager
```
This is incorrect. All all model attribute declarations must go into `Attrs` interfaces, otherwise you won't be able to fully construct this model (on the type-level obviously). Consult the original merge commit, because there may be more of this.

mattpap commented on bokehjs/src/coffee/models/sources/cds_view.ts:
```diff
@@ -76,36 +75,37 @@ export class CDSView extends Model {
   }
 
-  convert_indices_from_subset(indices) {
+  convert_indices_from_subset(indices: Number[]) {
```
Use lower case types for primitives. In fact, run `gulp scripts:tslint`, it will point out errors of this and other kinds.

mattpap commented on bokehjs/src/coffee/models/selections/selection.ts:
```diff
@@ -0,0 +1,115 @@
+import {merge} from "core/util/object"
+import {Glyph, GlyphView} from "models/glyphs/glyph"
+
+export class Selection extends Model {
```
You need to write type boilerplate for this and other new models (follow changes in PRs #7463 and 7467).

mattpap commented on bokehjs/src/coffee/core/selection_manager.ts:
```diff
@@ -20,39 +20,73 @@ export class SelectionManager extends HasProps {
-  clear(): void {
-    this.selector.clear()
-    this.source.selected = hittest.create_hit_test_result()
+  clear(rview: RendererView | undefined): void {
```
`clear(rview?: RenderView)`. This is why there are still compilation errors present. With the existing code, you allow empty value, but don't provide one, so you have to call this method with `clear(undefined)`.

mattpap commented on bokehjs/src/coffee/core/selection_manager.ts:
```diff
@@ -20,39 +20,73 @@ export class SelectionManager extends HasProps {
-    this.source.selected = hittest.create_hit_test_result()
+  clear(rview: RendererView | undefined): void {
+    this.source.selected.clear()
+    if(rview)
```
`if (view != null)` is the idiomatic approach. Also use single space between a keyword and opening parenthesis.

mattpap commented on bokehjs/src/coffee/core/selection_manager.ts:
```diff
@@ -20,39 +20,73 @@ export class SelectionManager extends HasProps {
+    if (glyph_renderer_views.length > 0) {
+      const hit_test_result = this.source.selection_policy.hit_test(geometry, renderer_views)
+      did_hit = did_hit || this.source.selection_policy.do_selection(hit_test_result, this.source, final, append)
+    }
```
This may be a subject for another PR, but this dichotomy doesn't look good. For example, adding another kind of renderer would be a pain. I suspect that usage of `model.type` is due to recursive imports? This furthermore suggests that this class shouldn't have any knowledge about specific kinds of renderers.

clairetang6 replied:
Yeah, I think the issue is that selection happens for groups of glyph renderers (that share a data source), rather than for each glyph renderer individually, so the glyph renderer can't be fully responsible for selection like the graph renderer can. There probably is a better solution, but at this point, leaving it for another PR seems ok.

philippjfr commented:
@clairetang6 Amazing, thanks for updating the tools! I'll try out the PR later today to confirm its all working but your changes look good.

bryevdv commented:
@clairetang6 two quick observations from testing so far: 

* ESC no longer seems to reset a selection (not sure if its due to the PR but I just noticed it)

* the `app/selection_histogram` example does not work, it reports the following:
   
   ```
   [Error] [bokeh] – "Unhandled ERROR reply to B0F7B71748EC47AA9B0F8352B0586F18: TypeError(\"'Selection' object is not subscriptable\",)"
   ```

clairetang6 commented:
@bryevdv, I fixed the selection_histogram example. For now, you need to set the initial `selected` property to an instance of Selection manually, because the default isn't being set on the python side. And now, instead of listening for changes to the data source's `selected` property, we can listen to changes to the selection's `indices` property. 

I haven't looked into the ESC behavior yet. I'm traveling this weekend, and don't think I'll be able to look into it until next week. I also have not really looked at updating examples, so there may still be some that don't work right now.

bryevdv commented:
> For now, you need to set the initial selected property to an instance of Selection manually, because the default isn't being set on the python side. And now, instead of listening for changes to the data source's selected property, we can listen to changes to the selection's indices property.

OK I think we will plan to release `0.12.14` before this PR then, to give time to get that sorted out first. Although we *can* update to the new syntax (and should, for purposes of demonstration) the old version should still function.

clairetang6 commented:
To fix the `TypeError(\"'Selection' object is not subscriptable\",)"`, I added a `__getitem__` to the Selection model. It feels kind of hacky, but I think it will be necessary to keep code that follows the selection_histogram example usage of `selection['1d']['indices']` working.

mattpap commented on bokeh/models/selections.py:
```diff
@@ -16,6 +19,24 @@ class Selection(Model):
+            logger.warn("['0d']['indices'] is deprecated. Use .line_indices instead.")
+            return None
+        elif key == '2d':
+            logger.warn("['2d']['indices'] is deprecated. Use .multiline_indices instead.")
```
Use `bokeh.util.deprecation`.

mattpap commented on bokehjs/src/coffee/models/selections/interaction_policy.ts:
```diff
@@ -13,6 +14,13 @@ export abstract class SelectionPolicy extends Model {
       source.selected.update(hit_test_result, final, append)
+
+      //new selection created in order for python-side change detection machinery
+      //to detect change in the source's selected property.
```
On a very remote side note, as you already used a period, make it a full sentence (start with a capital letter). It will look much more tidier than this. Also use single whitespace between comment indicator and comment content, for consistency with other code.

mattpap commented:
@bryevdv, I resolved the issues with this PR. Lets get this merged to avoid further merge conflicts.

clairetang6 commented:
@mattpap, thanks for the help.

bryevdv commented:
OK, I tested with relevant file and app examples and eveything seems to be in order. I will merge now, and we can get it in a dev build next week so users can kick it around a bit. 

@clairetang6 can you make any follow on issues, e.g. for updating examples, adding more docs, etc.  Thanks for your persistence and patience!
